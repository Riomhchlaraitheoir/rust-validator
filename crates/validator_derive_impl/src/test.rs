use quote::quote;
use syn::DeriveInput;
use crate::Input;

#[test]
fn struct_validator() {
    let input = quote! {
        struct SignupData {
            #[validator(email)]
            mail: String,
            #[validator(url)]
            site: String,
            #[validator(length(min = 1))]
            first_name: String,
            #[validator(range(18..))]
            age: u8,
            #[validator(elements)]
            dogs: Vec<Dog>
        }
    };

    let input: Input = syn::parse2(input).unwrap_or_else(|err| panic!("failed to parse input: {err}: {start:?} {end:?} ", start = err.span().start(), end = err.span().end()));
    let output = super::derive(input);
    let as_file = syn::parse_file(&output.to_string())
        .unwrap_or_else(|err| panic!("failed to parse outputted code: {err}\n{}", &output.to_string()));
    let formatted = prettyplease::unparse(&as_file);
    insta::assert_snapshot!(formatted)
}

#[test]
fn tuple_validator() {
    let input = quote! {
        struct SignupData(
            #[validator(email)]
            String,
            #[validator(url)]
            String,
            #[validator(length(min = 1))]
            String,
        );
    };

    let input: Input = syn::parse2(input).unwrap_or_else(|err| panic!("failed to parse input: {err}: {start:?} {end:?} ", start = err.span().start(), end = err.span().end()));
    let output = super::derive(input);
    let as_file = syn::parse_file(&output.to_string())
        .unwrap_or_else(|err| panic!("failed to parse outputted code: {err}\n{}", &output.to_string()));
    let formatted = prettyplease::unparse(&as_file);
    insta::assert_snapshot!(formatted)
}
#[test]
fn enum_validator() {
    let input = quote! {
        enum Request {
            Signup {
                #[validator(email)]
                mail: String,
                #[validator(url)]
                site: String,
                #[validator(length(min = 1))]
                first_name: String,
            },
            Login(#[validator(email)] String, #[validator(length(min = 8, max = 64))] String),
            Logout
        }
    };

    let input: Input = syn::parse2(input).unwrap_or_else(|err| panic!("failed to parse input: {err}: {start:?} {end:?} ", start = err.span().start(), end = err.span().end()));
    let output = super::derive(input);
    let as_file = syn::parse_file(&output.to_string())
        .unwrap_or_else(|err| panic!("failed to parse outputted code: {err}\n{}", &output.to_string()));
    let formatted = prettyplease::unparse(&as_file);
    insta::assert_snapshot!(formatted)
}


#[test]
fn list_validator() {
    let input = quote! {
        struct HasList {
            #[validator(elements)]
            list: Vec<Element>
        }
    };

    let input: Input = syn::parse2(input).unwrap_or_else(|err| panic!("failed to parse input: {err}: {start:?} {end:?} ", start = err.span().start(), end = err.span().end()));
    let output = super::derive(input);
    let as_file = syn::parse_file(&output.to_string())
        .unwrap_or_else(|err| panic!("failed to parse outputted code: {err}\n{}", &output.to_string()));
    let formatted = prettyplease::unparse(&as_file);
    insta::assert_snapshot!(formatted)
}


