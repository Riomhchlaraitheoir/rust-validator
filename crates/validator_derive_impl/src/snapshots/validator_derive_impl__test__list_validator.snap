---
source: crates/validator_derive_impl/src/test.rs
expression: formatted
---
#[derive(Debug, PartialEq, Clone)]
struct HasListValidationErrors {
    list: Option<
        ::validator::ElementsInvalid<
            <<<Vec<
                Element,
            > as ::validator::HasElements>::Element as ::validator::Validate>::Validator as ::validator::Validator<
                <Vec<Element> as ::validator::HasElements>::Element,
            >>::Error,
        >,
    >,
}
struct HasListValidator {
    list: ::validator::ElementsValidator<
        <<Vec<
            Element,
        > as ::validator::HasElements>::Element as ::validator::Validate>::Validator,
    >,
}
impl ::validator::Validator<HasList> for HasListValidator {
    type Error = HasListValidationErrors;
    fn validate(&self, HasList { list }: &HasList) -> Result<(), Self::Error> {
        let mut _valid = true;
        let validator = self;
        let error = HasListValidationErrors {
            list: {
                match validator.list.validate(list) {
                    Ok(()) => None,
                    Err(error) => {
                        _valid = false;
                        Some(error)
                    }
                }
            },
        };
        if _valid { Ok(()) } else { Err(error) }
    }
}
impl ::validator::Validate for HasList {
    type Validator = HasListValidator;
    fn validator() -> Self::Validator {
        HasListValidator {
            list: ::validator::ElementsValidator::new(
                <<Vec<
                    Element,
                > as ::validator::HasElements>::Element as ::validator::Validate>::validator(),
            ),
        }
    }
}
